integration-tests/sample_test.go

```diff
@@ -58,3 +63,147 @@ func TestAuth(t *testing.T) {
+	httpClient := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	restartSample(httpClient)
+
+	m := newModifier(modifySignatureHandler)
+	m.start()
+
+	defer m.stop()
+
+	payload := struct {
+		ModifyURL string `json:"modifyUrl"`
+	}{
+		ModifyURL: "http://127.0.0.1:" + fmt.Sprint(m.getPort()),
+	}
+
+	prreq, err := newRequest(fmt.Sprintf("http://%v:%v/session", options.proxyHost, options.proxyPort), "POST", payload)
+	if err != nil {
+		t.Fatalf("Error making request to the proxy: %v", err)
+	}
+
+	_, _, err = getResponse(prreq, httpClient)
+	if err != nil {
+		t.Fatalf("Error when receiving the proxy response: %v", err)
+	}
+
+	name := uuid.New().String()
+	userID := fmt.Sprintf("%v@example.com", name)
+	deviceName := "The device of " + name
+	pin := randPIN()
+
+	client := newSampleClient(options.sampleURL, httpClient)
+
+	authorizeRequestURL, err := client.authorize()
+	if err != nil {
+		t.Fatalf("Error making initial request to the home page: %v", err)
+	}
+
+	_, err = url.ParseRequestURI(string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Home page didn't respond with authorize url, error: %v, actual response: %v", err.Error(), string(authorizeRequestURL))
+	}
+
+	identity, err := register(httpClient, userID, deviceName, pin, string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Error registering: %v", err)
+	}
+
+	accessResponse, err := authenticate(httpClient, identity, userID, pin, string(authorizeRequestURL))
+	if err != nil {
+		t.Fatalf("Error authenticating: %v", err)
+	}
+
+	// Contains an URL, from which we can fetch the user info.
+	err = client.login(accessResponse.RedirectURL)
+	if err == nil {
+		t.Fatal("Expected error when the signature is invalid")
+	}
+
+	if !errors.Is(err, errInvalidLogin) {
+		t.Fatal("Expected invalid login error")
+	}
+
+	if !strings.Contains(err.Error(), "Invalid ID token") && !strings.Contains(err.Error(), "Signature validation failed.") {
+		t.Fatalf("Expected Invalid ID token error message")
+	}
+
+	// stop the proxy modifying session
+	prreq, err = newRequest(fmt.Sprintf("http://%v:%v/session", options.proxyHost, options.proxyPort), "DELETE", http.NoBody)
+	if err != nil {
+		t.Fatalf("Error when making stop session request to the proxy: %v", err)
+	}
+
+	_, _, err = getResponse(prreq, httpClient)
+	if err != nil {
+		t.Fatalf("Error on proxy response: %v", err)
+	}
+}
+
+func restartSample(client *http.Client) error {
+	restartSampleReq, err := newRequest("http://127.0.0.1:8081/restart?name="+options.sampleName, "POST", http.NoBody)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = getResponse(restartSampleReq, client)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(1 * time.Second)
+	return nil
+}
+
+func modifySignatureHandler(w http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
```

Comment:
Always defer after the error check. If there is an error this may panic.

Reply:
true..thanks..done